### PR TITLE
add solution

### PR DESCRIPTION
**Pull request by Victoria-Moiseienko**

### CODE REVIEW

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        return Storage.products.stream()
+                .filter(product -> product.getId().equals(productId))
+                .findFirst();
+    }
+
+    @Override
+    public Product update(Product product) {
+        Product currentProduct = get(product.getId()).get();
+        currentProduct.setName(product.getName());
+        currentProduct.setPrice(product.getPrice());
```
Comment (aleksandr-hrankin) [inline src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: а если у нас у product будет 40 полей, мы тоже их ручками будем обновлять?)

Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: :))

File: pom.xml [hunk 2]
```diff
@@ -8,5 +8,61 @@
...
+                        <source>${jdk.version}</source>
+                        <target>${jdk.version}</target>
+                        <encoding>${project.build.sourceEncoding}</encoding>
+                    </configuration>
+                </plugin>
+            </plugins>
+        </pluginManagement>
+    </build>
 
 </project>
```
Comment (dmytro-hryhoruk) [inline pom.xml]:
```suggestion
</project>

```

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java (hunk 1 above)
Comment (dmytro-hryhoruk) [reply src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: that would be easier to replace the old product in the storage using its id. You can do this using StreamApi

File: src/main/java/com/internet/shop/db/Storage.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import com.internet.shop.model.Product;
+import java.util.ArrayList;
+import java.util.List;
+
+public class Storage {
+    public static final List<Product> products = new ArrayList<>();
+    private static Long productId = 0L;
+
+    public static void addProduct(Product product) {
+        product.setId(++productId);
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/db/Storage.java]:
```suggestion
        product.setId(productId++);
```
It's okey to have the first index as zero

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java (hunk 1 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: Done

File: pom.xml (hunk 2 above)
Comment (Victoria-Moiseienko) [reply pom.xml]: Done

File: src/main/java/com/internet/shop/db/Storage.java (hunk 3 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/db/Storage.java]: Done

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+                .findFirst();
+    }
+
+    @Override
+    public Product update(Product product) {
+        getAllProducts().stream()
+                .filter(prod -> prod.getId().equals(product.getId()))
+                .forEach(prod -> prod = product);
+
+        return product;
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: let's pass the new Product Object with the same id and see if the list(storage) will change.

Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: done

File: pom.xml [hunk 5]
```diff
@@ -8,5 +8,61 @@
...
-</project>
\ No newline at end of file
+    <dependencies>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>4.12</version>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
```
Comment (KseniiaMakarova) [inline pom.xml]: let's remove this block
we don't use dependencies yet

File: src/main/java/com/internet/shop/Application.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    private static final Long APPLE_ID = 1L;
+
+    public static void main(String[] args) {
+        ProductService productService = (ProductService) injector.getInstance(ProductService.class);
+
+        productService.create(new Product("apple", 25.5));
+        productService.create(new Product("pear", 35.5));
+        productService.create(new Product("plum", 20));
+
+        productService.delete(APPLE_ID);
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/Application.java]: I think it is better to save products in variables and then call `apple.getId()` for example

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        IntStream.range(0, Storage.products.size())
+                .filter(i -> Storage.products.get(i).getId().equals(product.getId()))
+                .forEach(i -> Storage.products.set(i, product));
+
+        return product;
+    }
+
+    @Override
+    public boolean delete(Long productId) {
+        return Storage.products.remove(get(productId).orElseThrow());
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: let's use `removeIf()` here on `products`

File: src/main/java/com/internet/shop/factory/Factory.java [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+public class Factory {
+    private static ProductDao productDao;
+
+    public static ProductDao getProductDao() {
+        if (productDao == null) {
+            productDao = new ProductDaoImpl();
+        }
+        return productDao;
+    }
+}
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/factory/Factory.java]: do we need this class? this work is done by the Injector

File: src/main/java/com/internet/shop/lib/Dao.java [hunk 9]
```diff
@@ -0,0 +1,9 @@
+package com.internet.shop.lib;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+
+@Retention(RetentionPolicy.RUNTIME)
+public @interface Dao {
+}
+
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/lib/Dao.java]:
```suggestion
```

File: src/main/java/com/internet/shop/model/Order.java [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+
+import java.util.List;
+
+public class Order {
+    private Long id;
+    private List<Product> products;
+    private Long userId;
+
+    public Order(List<Product> products, Long userId) {
+        this.products = products;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/model/Order.java]: we will not receive `products` from outside
let's initialize `products` here as a new ArrayList

File: src/main/java/com/internet/shop/model/ShoppingCart.java [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+
+import java.util.List;
+
+public class ShoppingCart {
+    private Long id;
+    private List<Product> products;
+    private Long userId;
+
+    public ShoppingCart(List<Product> products, Long userId) {
+        this.products = products;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/model/ShoppingCart.java]: same

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 5 above)
Comment (Victoria-Moiseienko) [reply pom.xml]: Done

File: src/main/java/com/internet/shop/Application.java (hunk 6 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/Application.java]: Done

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java (hunk 7 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: Done

File: src/main/java/com/internet/shop/factory/Factory.java (hunk 8 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/factory/Factory.java]: Done

File: src/main/java/com/internet/shop/lib/Dao.java (hunk 9 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/lib/Dao.java]: Done

File: src/main/java/com/internet/shop/model/Order.java (hunk 10 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/model/Order.java]: Done

File: src/main/java/com/internet/shop/model/ShoppingCart.java (hunk 11 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/model/ShoppingCart.java]: Done

File: src/main/java/com/internet/shop/Application.java [hunk 12]
```diff
@@ -7,19 +7,21 @@
 
 public class Application {
     private static Injector injector = Injector.getInstance("com.internet.shop");
-    private static final Long PEAR_ID = 2L;
-    private static final Long APPLE_ID = 1L;
+
+    private static Product apple = new Product("apple", 25.5);
+    private static Product pear = new Product("pear", 35.5);
+    private static Product plum = new Product("plum", 20);
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/Application.java:13]: better make these variables local

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**